### PR TITLE
Record value expressions RFC

### PR DESCRIPTION
**Pull request by darh**

### CODE REVIEW

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+Library https://github.com/PaesslerAG/gval[gval] is used for parsing and evaluation of expression.
+This allows to reuse validators and sanitizers on front-end in JavaScript environment.
+
+== Value(s) expressions
+
+Value expressions allow dynamic calculation of record values from constants and other values.
+Only constants and values from current objects are supported.
+
+Corteza implements this functionality by allowing value expressions on any field type.
+When expression on the field input value on this field is ignored and expression is evaluated instead.
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: I can't make this out; can you please rewrite this one

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+----
+
+Expression assumes 2 other fields (firstName and lastName) and concatenates them with space between.
+Finally, we trim the result to ensure there result is not prefixed or suffixed with a space in case any of the two fields are empty.
+
+_Later upgrades will allow calculation from referenced records with dependency tracking and cascading updates._
+
+=== Field options
+
+ - formula
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: Use `*` for UL

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 3]
```diff
@@ -0,0 +1,212 @@
+include::../../variables.adoc[]
+
+= Module field expressions
+
+Module field Expressions enables support for arbitraty expressions on moduel fields and unlock:
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: ~moduel~ -- module

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+=== Field options
+
+ - formula
+
+=== Behaviour
+
+ - If an error occurs while executing formula field expression, error is appended to record value error set
+ - If expression on multi-value fields does not return slice of values, system raises error
+ - If expression on single-value fields does not return single value, system raises error
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: Use `*` and end each one with a `.`

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+|Name | Description
+
+| `old`
+| record object
+
+| `new`
+| record object
+
+| `<field-name>`
+| String or array of strings with new value.
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: I would prefer to see this in this form:

```
* `old`: the original record object.
* `new`: the new record object.
* `<field-name>`: String or array of strings with new value.
```

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+If field name collides with any of reserved variables (old, new, field, module) it can be accessed via `old.values.<field-name>`
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: I'd use a `[NOTE]` here.

Missing `.`

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+Sanitizers aid in record value processing and *cannot raise errors* but *can modify value*.
+Each module field can have one or more defined expression with accompanying description or explanation.
+
+.Example of length control
+[source]
+----
+length(value) > 5 ? substr(value, 0, 5) + "..." : value
+----
+
+Only one variable is exposed to the sanitizer
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: `.`

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+Sanitizer expression evaluation errors are logged.
+
+=== Available variables
+
+|===
+|Name | Description
+
+| `value`
+| sanitized value
+|===
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: Same as above (regarding list vs table)

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+| `value`
+| current value
+
+| `oldValue`
+| old value value
+
+| `values.<field-name>`
+| all values
+|===
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc]: Same as above; applies to all that follow

File: src/modules/ROOT/pages/rfc/record-value-expr/index.adoc [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+.Example of custom formatting
+[source]
+----
+capitalize(title)
+----
+
+== Expression functions (wip)
+
+|===
```
Comment (tjerman) [inline src/modules/ROOT/pages/rfc/record-value-expr/index.adoc:122]: This I would leave as it is (in a table); wrap the signature in a snippet

**Review by tjerman — CHANGES_REQUESTED** [review]
There are some grammar issues; I can go over it once you're done with it.

**Comment by darh** [comment]
Updated. 

Kindly asking @Fajfa to update the table at the end with the data he's working on for the frontend.

**Comment by Fajfa** [comment]
Table updated so it reflects the frontend help page tables

**Comment by tjerman** [comment]
Merged into https://github.com/cortezaproject/corteza-docs/commit/fa22e50de7dcc4b9f0d5dbdef66d1cea76171d54

**Pull request closed without merging**